File: TicTacToe/TicTacToe/GameBoard.vb
Public Class GameBoard

    Private _winner As IPlay

    Private _field(,) As Integer = { _
         {0, 0, 0}, _
         {0, 0, 0}, _
         {0, 0, 0} _
            }
    Private dic As Dictionary(Of Integer, Point)

    Public WriteOnly Property Winner As IPlay
        Set(value As IPlay)
            Me._winner = value
        End Set
    End Property

    Public Sub New()
        dic.Add(1, New Point(0, 0))
        dic.Add(2, New Point(0, 1))
        dic.Add(3, New Point(0, 2))
        dic.Add(4, New Point(1, 0))
        dic.Add(5, New Point(1, 1))
        dic.Add(6, New Point(1, 2))
        dic.Add(7, New Point(2, 0))
        dic.Add(8, New Point(2, 1))
        dic.Add(9, New Point(2, 2))
    End Sub


    Private Function GetCell(ByVal index As Integer) As Point
        If (dic.ContainsKey(index) = False) Then
            Return Nothing
        End If

        Return dic(index)

    End Function

    Public Function IsCellEmpty(ByVal i As Integer) As Boolean
        Dim pos As Point
        pos = GetCell(i)

        If (_field(pos.Y, pos.X) <> 0) Then
            Return False
        End If
        Return True

    End Function


    Public Sub UpdateCell(ByVal i As Integer, ByVal val As Integer)
        Dim pos As Point
        pos = GetCell(i)
        _field(pos.Y, pos.X) = val
    End Sub

    Public Function IsBoardEmpty()
        For y As Integer = 0 To 3
            For x As Integer = 0 To 3
                If _field(y, x) = 0 Then
                    Return False
                End If
            Next
        Next
        Return True
    End Function

    Public Function IsWin(ByRef h As HandType) As Boolean

        '
        For x As Integer = 0 To 3
            If (_field(0, x) = h And _
                _field(1, x) = h And _
                _field(2, x) = h) Then

                Return True
            End If
        Next

        '
        For y As Integer = 0 To 3
            If (_field(y, 0) = h And _
                _field(y, 1) = h And _
                _field(y, 2) = h) Then

                Return True
            End If
        Next

        '
        If (_field(0, 0) = h And _
            _field(1, 1) = h And _
            _field(2, 2) = h) Then

            Return True
        End If
        '

        If (_field(0, 2) = h And _
            _field(1, 1) = h And _
            _field(2, 0) = h And) Then _

            Return True
        End If

        '
        Return False

    End Function

    Public Sub GameStart()
        Console.WriteLine("/-/ game start! /-/")
    End Sub

    Public Sub GameFinish()
        Console.WriteLine("/-/ game was finished. /-/")
        If (_winner IsNot Nothing) Then
            _winner.WinnerInfo()
        Else
            Console.WriteLine("/-/ the game is draw. /-/")
        End If
    End Sub

    Public Function IsFinish() As Boolean
        ' TODO:
        Return False
    End Function




End Class

File: TicTacToe/TicTacToe/GameBoard.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace TicTacToe
{
    class GameBoard : IBoard
    {
        private IPlay _winner;
        private int[,] _field = {
                {0, 0, 0},
                {0, 0, 0},
                {0, 0, 0}
        };

        private Dictionary<Int32, Point> dic = new Dictionary<Int32, Point>();

        public GameBoard()
        {
            dic.Add(1, new Point(0, 0));
            dic.Add(2, new Point(0, 1));
            dic.Add(3, new Point(0, 2));
            dic.Add(4, new Point(1, 0));
            dic.Add(5, new Point(1, 1));
            dic.Add(6, new Point(1, 2));
            dic.Add(7, new Point(2, 0));
            dic.Add(8, new Point(2, 1));
            dic.Add(9, new Point(2, 2));
        }

        public void Display()
        {
            for (int y = 0; y < 3; ++y)
            {
                puts("+---+---+---+");
                for (int x = 0; x < 3; ++x)
                {
                    print("|");
                    PrintCell(_field[y, x]);
                }
                
                puts("|");
            }
            puts("+---+---+---+");
            puts("");
        }

        public void DisplayForSelectNumber()
        {
            puts("+---+---+---+");
            puts("| 1 | 2 | 3 |");
            puts("+---+---+---+");
            puts("| 4 | 5 | 6 |");
            puts("+---+---+---+");
            puts("| 7 | 8 | 9 |");
            puts("+---+---+---+");
            puts("");
        }


        public bool IsCellEmpty(int num)
        {
            Point pos = GetCell(num);

            if (_field[pos.Y, pos.X] != 0)
            {
                return false;
            }
            return true;
        }

        public bool IsAllCellEmpty()
        {
            for (int y = 0; y < 3; ++y)
            {
                for (int x = 0; x < 3; ++x)
                {
                    if (_field[y, x] == 0)
                    {
                        return false;
                    }
                }
            }
            return true;
        }

        public void GetCellValInfo(int num)
        {
            if (0 < num && num < 10)
            {
                puts(String.Format("select number[{0}] is {1}", num, GetCellStr(num)));
            }
        }

        public Point GetCell(int num)
        {
            return dic[num];
        }


        public void UpdateCell(Point pos, int num)
        {
            _field[pos.Y, pos.X] = num;
        }

        internal void startgame()
        {
            puts("### game start! ###");
        }

        public bool IsFinished { 
            get
            {
                return IsFinish();
            }
        }


        internal void endgame()
        {
            if (HasWinner)
            {
                _winner.WinInfo();
            }
            else
            {
                puts("game is draw.");
            }

            
            puts("### game is end. Otsukaresama Deshita. ###");
        }

        public bool IsWInner(IPlay p)
        {
            for (int y = 0; y < 3; ++y)
            {
                if (_field[y, 0] == p.GetHand() &&
                    _field[y, 1] == p.GetHand() &&
                    _field[y, 2] == p.GetHand())
                {
                    _winner = p;
                    return true;
                }

            }

            for (int x = 0; x < 3; ++x)
            {
                if (_field[0, x] == p.GetHand() &&
                    _field[1, x] == p.GetHand() &&
                    _field[2, x] == p.GetHand())
                {
                    _winner = p;
                    return true;
                }

            }

            if (_field[0, 0] == p.GetHand() &&
                _field[1, 1] == p.GetHand() &&
                _field[2, 2] == p.GetHand())
            {
                _winner = p;
                return true;
            }

            if (_field[0, 2] == p.GetHand() &&
                _field[1, 1] == p.GetHand() &&
                _field[2, 0] == p.GetHand())
            {
                _winner = p;
                return true;
            }

            return false;
        }

        public bool HasWinner
        {
            get
            {
                if (_winner == null)
                    return false;

                return true;
            }
        }

#region private methods
        private void print(string str)
        {
            Console.Write(str);
        }

        private void puts(string str)
        {
            Console.WriteLine(str);
        }

        private void PrintCell(int p)
        {
            print(GetCellStr(p));
        }

        private string GetCellStr(int p)
        {
            switch (p)
            {
                case 0:
                    return "   ";
                case 1:
                    return " o ";
                case 2:
                    return " x ";
                default:
                    return "   ";
            }
        }

        private bool IsFinish()
        {
            if (HasWinner || IsAllCellEmpty())
            {
                return true;
            }
            return false;
        }
#endregion
    }
}
